File: 백준/Gold/2458. 키 순서/키 순서.js
const input = require("fs")
  .readFileSync("/dev/stdin", "utf8")
  .trim()
  .split("\n")
  .map((line) => line.replace(/\r/, ""));

// 자기 노드를 포함해서 뒤에 있는놈 앞에 있는놈이 학생수와 같으면 등수를 알 수 있다
const solution = (input) => {
  const [studentN, lineN] = input[0].split(" ").map((v) => Number(v));
  let visited = Array(studentN + 1).fill(false);
  const relation = Array(studentN + 1).fill(0);
  const graph = {};

  for (let i = 1; i <= studentN; i++) {
    graph[i] = [];
  }

  for (let i = 1; i < input.length; i++) {
    const [n1, n2] = input[i].split(" ").map((v) => Number(v));
    graph[n1].push(n2);
  }

  const dfs = (node) => {
    relation[node] += 1;
    visited[node] = true;

    for (let nextNode of graph[node]) {
      if (visited[nextNode]) continue;
      dfs(nextNode);
    }
  };

  for (let i = 1; i <= studentN; i++) {
    dfs(i);
    relation[i] += visited.filter((v) => v).length - 1;
    visited = Array(studentN + 1).fill(false);
  }

  const answer = relation.filter((v) => v === studentN).length;
  console.log(answer);
};

solution(input);

File: 백준/Gold/2458. 키 순서/키 순서.py


'''
각 노드간의 모든거리를 구한다 -> 플로이드 워셜 알고리즘
거리를 알 수 있다 -> 연결되어있는 노드수가 student-1과 같으면 순위를 알 수 있다
'''

studentN, lineN = list(map(int, input().split()))
dist = [[float('inf')]*(studentN+1) for _ in range(studentN+1)]

for i in range(lineN):
  n1, n2 = list(map(int, input().split()))
  dist[n1][n2] = 1

for k in range(1, studentN+1):
  for i in range(1, studentN+1):
    for j in range(1, studentN+1):
      if dist[i][j] > dist[i][k] + dist[k][j]:
        dist[i][j] = dist[i][k] + dist[k][j]

answer = 0

for i in range(1, studentN+1):
  cnt = 0
  for j in range(1, studentN+1):
    if dist[i][j] != float('inf') or dist[j][i] != float('inf'):
      cnt += 1
  if cnt == studentN -1:
    answer +=1

print(answer)